File: frontend/src/api/minutes.ts
import type { User } from 'firebase/auth';

import { fetchWithAuth, readErrorDetail } from './http';

const defaultMinutesError = '議事録の生成に失敗しました。時間をおいて再試行してください。';

type MinutesFromTextResponse = {
  minutes?: {
    body?: unknown;
  };
};

function getReadableMinutesError(detail: string, status: number) {
  if (detail === 'text is required') {
    return '文字起こしを入力してください。';
  }

  if (detail === 'text must be 50000 characters or less') {
    return '文字起こしは50,000文字以内にしてください。';
  }

  if (
    status === 401 ||
    detail === 'authorization bearer token is required' ||
    detail === 'invalid firebase id token'
  ) {
    return 'ログイン状態を確認してください。';
  }

  if (detail === 'failed to generate minutes') {
    return defaultMinutesError;
  }

  return defaultMinutesError;
}

export async function generateMinutesFromText(user: User, text: string) {
  let response: Response;
  try {
    response = await fetchWithAuth(user, '/api/minutes/from-text', {
      method: 'POST',
      body: JSON.stringify({ text }),
    });
  } catch (err) {
    throw err instanceof Error ? err : new Error('バックエンドに接続できません。');
  }

  if (!response.ok) {
    const detail = await readErrorDetail(response);
    throw new Error(getReadableMinutesError(detail, response.status));
  }

  const payload = (await response.json()) as MinutesFromTextResponse;
  const body = payload.minutes?.body;
  if (typeof body !== 'string') {
    throw new Error('議事録生成APIのレスポンスを読み取れませんでした。');
  }

  return body;
}

File: backend/app/api/minutes.py
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from starlette.concurrency import run_in_threadpool

from app.core.auth import AuthenticatedUser, verify_firebase_user
from app.services.minutes_service import (
    MAX_TEXT_LENGTH,
    MinutesGenerationError,
    generate_minutes_from_text,
)


router = APIRouter()


class MinutesFromTextRequest(BaseModel):
    text: str | None = None


class MinutesBody(BaseModel):
    body: str


class MinutesFromTextResponse(BaseModel):
    minutes: MinutesBody


@router.post("/from-text", response_model=MinutesFromTextResponse)
async def minutes_from_text(
    request: MinutesFromTextRequest | None = None,
    _user: AuthenticatedUser = Depends(verify_firebase_user),
) -> MinutesFromTextResponse:
    text = request.text if request else ""
    text = text or ""
    if not text.strip():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="text is required",
        )

    if len(text) > MAX_TEXT_LENGTH:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="text must be 50000 characters or less",
        )

    try:
        body = await run_in_threadpool(generate_minutes_from_text, text.strip())
    except MinutesGenerationError as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="failed to generate minutes",
        ) from exc

    return MinutesFromTextResponse(minutes=MinutesBody(body=body))
